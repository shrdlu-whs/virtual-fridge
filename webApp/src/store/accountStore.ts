/* eslint-disable no-debugger */
import { makeObservable, action, computed, observable } from "mobx";
import { Auth } from "aws-amplify";
import { RootStore } from "./rootStore";
import {
  IRegisterForm,
  ISignInForm,
  IForgotPasswordForm,
  IForgotPasswordSubmitForm,
  IUserInfo,
} from "../interfaces/account.interface";
import axios, { Canceler, CancelToken, CancelTokenSource, CancelTokenStatic } from "axios";


// Amplify: https://docs.amplify.aws/lib/auth/getting-started/q/platform/js

export class AccountStore {
  @observable userInfo: IUserInfo | null = null;
  @observable forgotUsername: string | undefined = undefined;

  // cancelToken: CancelTokenSource = axios.CancelToken.source()
  cancelRequest: Canceler | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  generateCancelToken() {
    return new axios.CancelToken((c) => {
      this.cancelRequest = c;
    })
  }

  @action
  async register(register: IRegisterForm) {
    try {
      await Auth.signUp({
        username: register.username,
        password: register.password,
        attributes: {
          email: register.email,
        },
      });
      this.rootStore.routeStore.redirect("/signIn");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "register",
        "error",
        true,
        error
      );
    }
  }

  @action
  async signIn(signIn: ISignInForm) {
    try {
      this.rootStore.uiStateStore.setSubmitting(true);
      await Auth.signIn(signIn.username, signIn.password);
      await this.fetchUserInfo();
      this.rootStore.routeStore.redirect("/products");
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signIn",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.uiStateStore.setSubmitting(false);
    }
  }

  @action
  async signOut() {
    try {
      await Auth.signOut();
      // this.userInfo = null; // If the site refresh, fetchUserInfo() will be executed and the userInfo will set anyway
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "signOut",
        "error",
        true,
        error
      );
    } finally {
      this.rootStore.routeStore.redirect("/signIn");
    }
  }

  @action
  async forgotPassword({ username }: IForgotPasswordForm) {
    try {
      await Auth.forgotPassword(username);
      this.setForgotUsername(username);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async forgotPasswordSubmit({
    username,
    code,
    password,
  }: IForgotPasswordSubmitForm) {
    try {
      await Auth.forgotPasswordSubmit(username, code, password);
      this.rootStore.routeStore.redirect("/signIn");
      this.setForgotUsername(undefined);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "forgot_password_submit",
        "error",
        true,
        error
      );
    }
  }

  @action
  async changePassword(oldPassword: string, newPassword: string) {
    try {
      const user = await Auth.currentAuthenticatedUser();
      await Auth.changePassword(user, oldPassword, newPassword);
    } catch (error) {
      this.rootStore.notificationStore.createNotification(
        "change_password",
        "error",
        true,
        error
      );
    }
  }

  @action
  async fetchUserInfo() {
    if (!this.rootStore.uiStateStore.fetchingUser) {
      try {
        this.rootStore.uiStateStore.setFetchingUser(true);
        const userInfo = (await Auth.currentUserInfo()) as IUserInfo | null;
        // const userInfo: IUserInfo = {username: "Test", attributes: {email: "test@mail.com", email_verified: "true", sub: "" }};
        this.setUserInfo(userInfo);
      } catch (error) {
        this.setUserInfo(null);
        this.rootStore.notificationStore.createNotification(
          "fetch_user",
          "error",
          true,
          error
        );
      } finally {
        this.rootStore.uiStateStore.setFetchingUser(false);
        // Its initiating, also when the an error occured => the user returns to login page
        if(!this.rootStore.uiStateStore.isInitiating) {
          this.rootStore.uiStateStore.setInitiated(true);
        }
      }
    }
  }

  @action
  setUserInfo(userInfo: IUserInfo | null) {
    this.userInfo = userInfo;
  }

  @action
  setForgotUsername(username: string | undefined) {
    this.forgotUsername = username;
  }
}
