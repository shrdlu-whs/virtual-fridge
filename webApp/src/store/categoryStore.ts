/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { getCategories } from "../services/category2.service";

export class CategoryStore {
  @observable categories: IApiCategory[] = [];

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getCategories() {
    try {
      const categories = (await getCategories()).category;
      this.setCategories(categories);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_categories", "error", true, error);
      }
    }
  }

  @action
  setCategories(categories: IApiCategory[]) {
    this.categories = categories;
  }
}
