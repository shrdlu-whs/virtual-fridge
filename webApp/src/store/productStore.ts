/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { IApiProduct } from "../interfaces/api/api-product.interface";
import { IIronStockProduct } from "../interfaces/ironStockProduct.interface";
import { getProducts, getProduct, addProduct, updateProduct, deleteProduct } from "../services/product2.service";
import { IApiCategory } from "../interfaces/api/api-category.interface";
import { INutritionValue, IProductItem, IProduct } from "../interfaces/froms/addProductForm.interface";
import { EMPTY_OBJECT } from "mobx/dist/internal";

export class ProductStore {

  nutritionValueEmpty: INutritionValue = {
    calories: 0,
    carbohydrates: 0,
    fat: 0,
    fiber: 0,
    nutritionScore: "A",
    protein: 0,
    salt: 0,
    saturatedFat: 0,
    sugar: 0
  };

  @observable activeProduct: IApiProduct | undefined = undefined;
  @observable products: IApiProduct[] = [];
  @observable ironStockProducts: IIronStockProduct[] = [];
  @observable missingIronStockProducts: IIronStockProduct[] = [];
  @observable productItems: IProductItem[] = [];

  @observable addProductFormData: IProduct = {
    name: "",
    iconPath: "",
    category: [],
    totalQuantity: 1,
    minQuantity: -1,
    unit: "",
    items: this.productItems,
    nutritionalValue: this.nutritionValueEmpty
  };

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getProducts() {
    try {
      // const products = (await getProducts(this.rootStore.accountStore.generateCancelToken())).data;
      this.rootStore.uiStateStore.setLoading(true);
      const products = (await getProducts()).products;
      this.setProducts(products);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async getProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      // const product = (await getProduct(id, this.rootStore.accountStore.generateCancelToken())).data;
      const product = (await getProduct(id)).product;
      this.setActiveProduct(product);
      this.updateCachedProducts(product);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_product", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      product = (await addProduct(product)).createProduct;
      this.setProducts(this.products.concat([product]));
    } catch (error) {
      console.log(error)
      this.rootStore.notificationStore.createNotification("add_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateProduct(product: IApiProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await updateProduct(product);
      if (product.id) this.updateCachedProducts(product);
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async deleteProduct(id: number) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      await deleteProduct(id);
      const index = this.products.findIndex((product) => product.id === id);
      if (index !== -1) {
        runInAction(() => {
          this.products.splice(index, 1);
        });
      }
    } catch (error) {
      this.rootStore.notificationStore.createNotification("delete_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedProducts(product: IApiProduct) {
    // Update cached products
    const index = this.products.findIndex((sProduct) => sProduct.id === product.id);
    if (index !== -1) {
      this.products.splice(index, 1, product);
    } else {
      this.products.push(product);
    }
  }

  async getIronStockProducts() {
    try {
      const products = (await getProducts()).products;
      this.rootStore.uiStateStore.setLoading(true);
      this.setProducts(products);
      const ironStockProducts = products.map(
        (product): IIronStockProduct => {
          return { id: product.id!, name: product.name, stock: product.totalQuantity, ironStock: product.minQuantity };
        }
      );
      this.setIronStockProducts(ironStockProducts.filter((product) => product.ironStock >= 0));
      this.setMissingIronStockProducts(ironStockProducts.filter((product) => product.ironStock < 0));
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_products", "error", true, error);
      }
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async addIronStockProduct(ironStockProduct: IIronStockProduct) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      const cachedProduct = this.products.find((sProduct) => sProduct.id === ironStockProduct.id)!;
      cachedProduct.minQuantity = ironStockProduct.ironStock;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.concat([ironStockProduct]));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  async updateIronStockProducts(ids: number[]) {
    try {
      this.rootStore.uiStateStore.setLoading(true);
      for (const id of ids) {
        const product = this.ironStockProducts.find((product) => product.id === id);
        if (product) {
          const cachedProduct = this.products.find((sProduct) => sProduct.id === product.id)!;
          cachedProduct.minQuantity = product.ironStock;
          await updateProduct(cachedProduct);
        }
      }
    } catch (error) {
      // Rollback ???
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    } finally {
      this.rootStore.uiStateStore.setLoading(false);
    }
  }

  @action
  updateCachedIronStockProduct(id: number, ironStock: number) {
    const ironStockProduct = this.ironStockProducts.find((product) => product.id === id);
    if (ironStockProduct) {
      ironStockProduct.ironStock = ironStock;
    }
  }

  @action
  async deleteIronStockProduct(id: number) {
    try {
      const cachedProduct = this.products.find((sProduct) => sProduct.id === id)!;
      cachedProduct.minQuantity = -1;
      await updateProduct(cachedProduct);
      this.setIronStockProducts(this.ironStockProducts.filter((product) => product.id !== id));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_product", "error", true, error);
    }
  }

  @action
  setActiveProduct(product: IApiProduct) {
    this.activeProduct = product;
  }

  @action
  setProducts(products: IApiProduct[]) {
    this.products = products;
  }

  @action
  setIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.ironStockProducts = ironStockProducts;
  }

  @action
  setMissingIronStockProducts(ironStockProducts: IIronStockProduct[]) {
    this.missingIronStockProducts = ironStockProducts;
  }

  @action
  setProductItems(items: IProductItem[]) {
    this.productItems = items;
  }

  @action
  setAddProductFormData(data: IProduct) {
    this.addProductFormData = data;
  }

  @action
  resetAddProductFormData = () => {
    this.setProductItems([]);
    this.setAddProductFormData({
      name: "",
      iconPath: "",
      category: [],
      totalQuantity: 1,
      minQuantity: -1,
      unit: "",
      items: this.productItems,
      nutritionalValue: this.nutritionValueEmpty
    });
  };
}
