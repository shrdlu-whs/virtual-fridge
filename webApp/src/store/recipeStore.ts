/* eslint-disable no-debugger */
import { makeObservable, action, observable, runInAction } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist";
import { getRecipe, getRecipes, addRecipe, updateRecipe, deleteRecipe } from "../services/recipe2.service";
import { IApiRecipe } from "../interfaces/api/api-recipe.interface";
import { IApiItem } from "../interfaces/api/api-item.interface";
import { IApiRecipeItem } from "../interfaces/api/api-recipeItem.interface";
import { IRecipe } from "../interfaces/froms/addRecipeForm.interface";
import { IRecipeItem } from "../interfaces/froms/addRecipeForm.interface";
export class RecipeStore {
  @observable activeRecipe: IApiRecipe | undefined = undefined;
  @observable recipes: IApiRecipe[] = [];

  //#region Add recipe form data props
  @observable recipeItems: IRecipeItem[] = [];
  @observable addRecipeFormData: IRecipe = {
    name: "",
    shortDescription: "",
    instructions: "",
    hyperlink: "https://www.someimage.net",
    expectedTime: "00:05:00",
    difficulty: "easy",
    recipeItems: this.recipeItems,
  };
  //#endregion

  constructor(private rootStore: RootStore) {
    makeObservable(this);
  }

  async getRecipes() {
    try {
      const recipes = (await getRecipes()).recipes;
      this.setRecipes(recipes);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_recipes", "error", true, error);
      }
    }
  }

  async getRecipe(id: number) {
    try {
      const recipe = (await getRecipe(id)).recipe;
      this.setActiveRecipe(recipe);
      this.updateCachedRecipes(recipe);
    } catch (error) {
      if (error.message) {
        this.rootStore.notificationStore.createNotification("get_recipe", "error", true, error);
      }
    }
  }

  async addRecipe(recipe: IApiRecipe) {
    try {
      recipe = (await addRecipe(recipe)).createRecipe;
      this.setRecipes(this.recipes.concat([recipe]));
    } catch (error) {
      this.rootStore.notificationStore.createNotification("add_recipe", "error", true, error);
    }
  }

  async updateRecipe(recipe: IApiRecipe) {
    try {
      await updateRecipe(recipe);
      if (recipe.id) this.updateCachedRecipes(recipe);
    } catch (error) {
      this.rootStore.notificationStore.createNotification("update_recipe", "error", true, error);
    }
  }

  async deleteRecipe(id: number) {
    try {
      await deleteRecipe(id);
      const index = this.recipes.findIndex((product) => product.id === id);
      if (index !== -1) {
        runInAction(() => {
          this.recipes.splice(index, 1);
        });
      }
    } catch (error) {
      this.rootStore.notificationStore.createNotification("delete_recipe", "error", true, error);
    }
  }

  updateCachedRecipes(recipe: IApiRecipe) {
    // Update cached products
    const index = this.recipes.findIndex((sRecipe) => sRecipe.id === recipe.id);
    if (index !== -1) {
      this.recipes.splice(index, 1, recipe);
    } else {
      this.recipes.push(recipe);
    }
  }

  @action
  setActiveRecipe(recipe: IApiRecipe) {
    this.activeRecipe = recipe;
  }

  @action
  setRecipes(recipes: IApiRecipe[]) {
    this.recipes = recipes;
  }

  @action
  setRecipeItems(items: IRecipeItem[]) {
    this.recipeItems = items;
  }

  @action
  setAddRecipeFormData(data: IRecipe) {
    this.addRecipeFormData = data;
  }

  @action
  resetAddRecipeFormData = () => {
    this.setRecipeItems([]);
    this.setAddRecipeFormData({
      name: "",
      shortDescription: "",
      instructions: "",
      hyperlink: "https://www.someimage.net",
      expectedTime: "00:05:00",
      difficulty: "easy",
      recipeItems: this.recipeItems,
    });
  };
}
