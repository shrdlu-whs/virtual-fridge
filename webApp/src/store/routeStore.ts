import { makeObservable, action, observable } from "mobx";
import { fromStream } from "mobx-utils";
import { RootStore } from "./rootStore";
import { Redirect } from "react-router-dom"

export class RouteStore {
  @observable redirectedTo: string | undefined = undefined;

  constructor(private rootStore: RootStore) {
    makeObservable(this)
  }

  @action
  redirect(path: string, tokenExpired: boolean = false) {
    if (this.redirectedTo !== path) {
      if (tokenExpired) {
        this.rootStore.notificationStore.createNotification(
          "token_expired",
          "error",
          true
        );
      }
      this.redirectedTo = path;
    }
  }

  @action
  resetRedirectedTo() {
    this.redirectedTo = undefined;
  }
}
