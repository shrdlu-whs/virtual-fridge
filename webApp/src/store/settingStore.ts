/* eslint-disable no-debugger */
import { makeObservable, action, observable } from "mobx";
import { RootStore } from "./rootStore";
import { persist } from "mobx-persist"

export class SettingStore {

  constructor(private rootStore: RootStore) {
    // makeObservable(this)
  }

}
