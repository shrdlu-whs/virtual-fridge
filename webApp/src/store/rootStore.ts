import { UiStateStore } from "./uiStateStore";
import { AccountStore } from "./accountStore";
import { RouteStore } from "./routeStore";
import { SettingStore } from "./settingStore";
import { NotificationStore } from "./notificationStore"
import { ProductStore } from "./productStore"
import { RecipeStore } from "./recipeStore"
import { ShoppingListStore } from "./shoppinglistStore"
import { CategoryStore } from "./categoryStore";


export class RootStore {
  uiStateStore: UiStateStore;
  accountStore: AccountStore;
  routeStore: RouteStore;
  notificationStore: NotificationStore;
  settingStore: SettingStore;
  productStore: ProductStore;
  recipeStore: RecipeStore;
  shoppingListStore: ShoppingListStore;
  categoryStore: CategoryStore;

  constructor() {
    this.uiStateStore = new UiStateStore(this);
    this.accountStore = new AccountStore(this);
    this.routeStore = new RouteStore(this);
    this.notificationStore = new NotificationStore(this);
    this.settingStore = new SettingStore(this);
    this.productStore = new ProductStore(this);
    this.recipeStore = new RecipeStore(this);
    this.shoppingListStore = new ShoppingListStore(this);
    this.categoryStore = new CategoryStore(this);
  }
}
