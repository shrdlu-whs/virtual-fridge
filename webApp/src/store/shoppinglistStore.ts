import { makeObservable, action, observable, runInAction } from "mobx";
import { persist } from "mobx-persist";
import { IApiShoppingList, IUpdateShoppingList } from '../interfaces/api/api-shoppinglist.interface';

import { IAddShoppingList } from '../interfaces/api/api-shoppinglist.interface';
import { RootStore } from "./rootStore";
import { getShoppingLists, getShoppingList, addShoppingList, updateShoppingList, deleteShoppingList } from '../services/shoppingList2.service'
import { IAddShoppinglistItem } from "../interfaces/api/api-shoppinglistItem.interface";

export class ShoppingListStore {
    @observable activeShoppinglist: IApiShoppingList | undefined = undefined;
    @observable shoppingLists: IApiShoppingList[] = [];
    @observable shoppingList: IAddShoppingList[] = [];
 
    @observable shoppingListsItems: IAddShoppinglistItem[] = [];
    @observable addShoppingListFormData: IAddShoppingList = {
        name: "",
        shoppingListItems: this.shoppingListsItems,
        created: new Date()
    };
    constructor(private rootStore: RootStore) {
        makeObservable(this);
    }
    async getShoppinglists() {
        try {
            this.rootStore.uiStateStore.setLoading(true);
            const shoppinglists = (await getShoppingLists()).shoppingLists;
            this.setShoppingLists(shoppinglists);
            console.log(shoppinglists);
            console.log(this.addShoppingListFormData);
        } catch (error) {
            this.rootStore.notificationStore.createNotification("get_shoppinglists", "error", true, error);
        } finally {
            this.rootStore.uiStateStore.setLoading(false);
        }
    }
    async getShoppinglist(id: number) {
        try {
            this.rootStore.uiStateStore.setLoading(true);
            const shoppinglist = (await getShoppingList(id)).shoppingList;
            this.setActiveShoppingList(shoppinglist);
            this.updateCachedShoppingLists(shoppinglist);
            console.log(this.activeShoppinglist);
        } catch (error) {
            this.rootStore.notificationStore.createNotification("get_shoppinglist", "error", true, error);
        } finally {
            this.rootStore.uiStateStore.setLoading(false);
        }
    }
    async addShoppingList(shoppinglist: IAddShoppingList) {
        try {
            this.rootStore.uiStateStore.setLoading(true);
            shoppinglist = (await addShoppingList(shoppinglist)).createShoppingList;
            this.setShoppingList(this.shoppingList.concat([shoppinglist]));
        } catch (error) {
            this.rootStore.notificationStore.createNotification("add_shoppinglist", "error", true, error);
        } finally {
            this.rootStore.uiStateStore.setLoading(false);
        }
    }
    
    async updateShoppingList(shoppinglist: IUpdateShoppingList) {
        try {
            this.rootStore.uiStateStore.setLoading(true);
            await updateShoppingList(shoppinglist);
        } catch (error) {
            this.rootStore.notificationStore.createNotification("update_shoppinglist", "error", true, error);
        } finally {
            this.rootStore.uiStateStore.setLoading(false);
        }

    }

    async deleteShoppingList(id: number) {
        try {
            this.rootStore.uiStateStore.setLoading(true);
            await deleteShoppingList(id);
            const index = this.shoppingLists.findIndex((shoppingList) => shoppingList.id === id);
            if (index !== -1) {
                runInAction(() => {
                    this.shoppingLists.splice(index, 1);
                })
            }
        } catch (error) {
            this.rootStore.notificationStore.createNotification("delete_shoppinglist", "error", true, error);
        } finally {
            this.rootStore.uiStateStore.setLoading(false);
        }
    }

    @action
    updateCachedShoppingLists(shoppinglist: IApiShoppingList) {
        const index = this.shoppingLists.findIndex((sShoppingList) => sShoppingList.id === shoppinglist.id);
        if (index !== -1) {
            this.shoppingLists.splice(index, 1, shoppinglist);
        } else {
            this.shoppingLists.push(shoppinglist);
        }
    }

    @action
    updateCachedShoppingListItem(id: number, quantity: number) {
        const shoppingListItem = this.activeShoppinglist?.shoppingListItems.find((i) =>i.id === id);
       console.log(shoppingListItem);
       console.log(quantity);
        if(shoppingListItem) {
            shoppingListItem!.quantity = quantity;
            console.log(shoppingListItem);
        }
        updateShoppingList(this.activeShoppinglist!);
    }

    @action
    setShoppingList(shoppinglist: IAddShoppingList[]) {
        this.shoppingList = shoppinglist;
    }
    @action
    setShoppingLists(shoppinglists: IApiShoppingList[]) {
        this.shoppingLists = shoppinglists;
    }

    @action
    setActiveShoppingList(shoppinglist: IApiShoppingList) {
        this.activeShoppinglist = shoppinglist;
    }

    @action
    setShoppingListItems(items: IAddShoppinglistItem[]) {
        this.shoppingListsItems = items;
    }
  
    @action
    resetAddShoppingListFormData() {
        this.setShoppingListItems([]);
        this.setAddShoppingListFormData({
            name: "",
            shoppingListItems: this.shoppingListsItems,
            created: new Date()
        });
    }

    @action 
    setAddShoppingListFormData(data: IAddShoppingList) {
        this.addShoppingListFormData = data;
    }
}