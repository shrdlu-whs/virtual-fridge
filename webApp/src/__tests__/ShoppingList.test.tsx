import React from "react";
import IronStock from '../components/iron-stock-content/ironStockContainer'
import { fireEvent, render, screen } from '@testing-library/react'; 
import ShoppingList from '../components/grocerylist-content/shoppinglists/shoppingListsContainer'
import { RootStore } from "../store/rootStore";
import { Provider } from "mobx-react";
import { createMemoryHistory } from 'history'
import { Router } from 'react-router-dom'
import userEvent from '@testing-library/user-event'
import { IAddShoppingList, IApiShoppingList } from './../interfaces/api/api-shoppinglist.interface'
import GroceryList from '../components/grocerylist-content/groceryListContainer'


const shoppinglist = require('../components/grocerylist-content/shoppinglists/shoppingListsContainer')
test ('Register exists', () => {
    expect(typeof shoppinglist).toEqual('object');  
});

test('render about link', () => {
    const rootStore: RootStore = new RootStore();
    const history = createMemoryHistory();
    render(
        <Provider {...rootStore}>
            <ShoppingList />
         </Provider>)
    const addButton = screen.getByRole('button', {name: 'add'});
    const deleteButton = screen.getByRole('button', {name: 'delete'});
    const spanName = screen.getByRole('button', {name: 'Name'});
    const table = screen.getByRole('table', {name: /title\.grocery\_list\_table/i});
    expect(deleteButton).toBeInTheDocument();
    expect(addButton).toBeInTheDocument();
    fireEvent.click(addButton);
    expect(open).toBeTruthy();
    expect(spanName).toBeInTheDocument();    
    expect(table).toBeInTheDocument();    

})

test('test buttons onClick', () => {
    const rootStore: RootStore = new RootStore();
    const history = createMemoryHistory();
    render(
        <Provider {...rootStore}>
            <ShoppingList />
         </Provider>)
    const addButton = screen.getByRole('button', {name: 'add'}); 
    fireEvent.click(addButton);
    console.log(open);
    expect(open).toBeTruthy();  
})

test('test object of shoppinglist', () => { 
    const shoppinglist: IAddShoppingList = {
        name: "Shopping",
        items: [],
        created: new Date("2021-02-14T15:27:16.098Z")
      };
      expect(shoppinglist).toMatchSnapshot();
})


/*test('rendering Formik form',() => {
    const rootStore: RootStore = new RootStore();
    render(
        <Provider {...rootStore}>
            <ShoppingList />
         </Provider>)
    const label = screen.getByLabelText('Name')
    userEvent.type(label, 'Shopping');

})*/

