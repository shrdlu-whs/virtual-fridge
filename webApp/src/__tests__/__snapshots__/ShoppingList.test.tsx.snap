// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test object of shoppinglist 1`] = `
Object {
  "created": 2021-02-14T15:27:16.098Z,
  "items": Array [],
  "name": "Shopping",
}
`;
